### PR TITLE
Allow unix-abstract uri's to use local_security_connector

### PR DESCRIPTION
**Pull request by beshaya**
@markdroth

### CODE REVIEW

**Review by markdroth — COMMENTED** [review]
This looks good to me, but I'd like @jiangtaoli2016 or @yihuazhang to confirm.

Also, we should probably have a test for this.

File: src/core/lib/security/security_connector/local/local_security_connector.cc [hunk 1]
```diff
@@ -257,7 +258,9 @@ grpc_local_channel_security_connector_create(
   const char* server_uri_str = grpc_channel_arg_get_string(server_uri_arg);
   if (creds->connect_type() == UDS &&
       strncmp(GRPC_UDS_URI_PATTERN, server_uri_str,
-              strlen(GRPC_UDS_URI_PATTERN)) != 0) {
+              sizeof(GRPC_UDS_URI_PATTERN) - 1) != 0 &&
+      strncmp(GRPC_UDS_URI_PATTERN, server_uri_str,
```
Comment (yihuazhang) [inline src/core/lib/security/security_connector/local/local_security_connector.cc:262]: s/GRPC_UDS_URI_PATTERN/GRPC_UDS_ABSTRACT_URI_PATTERN

**Review by yihuazhang — COMMENTED** [review]
Based on https://github.com/grpc/grpc/blob/master/doc/naming.md, unix abstract namespace does not have permissions. So I do not think we should treat it in the same way as the conventional unix sockets (`unix:`) by assigning `PRIVACY_AND_INTEGRITY` to it. It is closer to local TCP from security-wise.

File: src/core/lib/security/security_connector/local/local_security_connector.cc (hunk 1 above)
Comment (beshaya) [reply src/core/lib/security/security_connector/local/local_security_connector.cc]: Food catch. Fixed.

**Comment by beshaya** [comment]
> Based on https://github.com/grpc/grpc/blob/master/doc/naming.md, unix abstract namespace does not have permissions. So I do not think we should treat it in the same way as the conventional unix sockets (`unix:`) by assigning `PRIVACY_AND_INTEGRITY` to it. It is closer to local TCP from security-wise.

> It looks like TSI_PRIVACY_AND_INTEGRITY is assigned to both tcp and uds currently:
```
  // TODO(yihuazhang): Set security level of local TCP to TSI_SECURITY_NONE.
  const char* security_level =
      tsi_security_level_to_string(TSI_PRIVACY_AND_INTEGRITY);
  tsi_result result = tsi_construct_string_peer_property_from_cstring(
      TSI_SECURITY_LEVEL_PEER_PROPERTY, security_level,
      &peer.properties[peer.property_count]);
```

I'm curious about the logic of giving conventional unix sockets a higher security level. The filesystem makes it possible to enforce permissions, but (to my knowledge) gRPC doesn't check that the permissions aren't 777. In our use case, we are verifying the UID of connecting processes to our unix abstract socket, which is effectively the same as the filesystem check for conventional sockets.

What do you recommend our team do?

**Comment by yihuazhang** [comment]
> > Based on https://github.com/grpc/grpc/blob/master/doc/naming.md, unix abstract namespace does not have permissions. So I do not think we should treat it in the same way as the conventional unix sockets (`unix:`) by assigning `PRIVACY_AND_INTEGRITY` to it. It is closer to local TCP from security-wise.
> 
> > It looks like TSI_PRIVACY_AND_INTEGRITY is assigned to both tcp and uds currently:

Yes, we need to change the security level for local TCP to INSECURE. It was a technical debt. 
> 
> ```
>   // TODO(yihuazhang): Set security level of local TCP to TSI_SECURITY_NONE.
>   const char* security_level =
>       tsi_security_level_to_string(TSI_PRIVACY_AND_INTEGRITY);
>   tsi_result result = tsi_construct_string_peer_property_from_cstring(
>       TSI_SECURITY_LEVEL_PEER_PROPERTY, security_level,
>       &peer.properties[peer.property_count]);
> ```
> 
> I'm curious about the logic of giving conventional unix sockets a higher security level. The filesystem makes it possible to enforce permissions, but (to my knowledge) gRPC doesn't check that the permissions aren't 777. In our use case, we are verifying the UID of connecting processes to our unix abstract socket, which is effectively the same as the filesystem check for conventional sockets.
> 
> What do you recommend our team do?

You are right that gRPC does not perform any permission checks. It is up to an application to correctly set the file permission. For abstracted unix socket, a user does not have this capability. I would recommend creating a new enum value for `UDS_ABSTRACT`, assign `GRPC_SECURITY_NONE` to it, and have a separate code path for it in `local_security_connector.cc`.

**Comment by beshaya** [comment]
Is LocalCredentials with GRPC_SECURITY_NONE meaningfully different from
grpc::InsecureChannelCredentials?

**Comment by yihuazhang** [comment]
> Is LocalCredentials with GRPC_SECURITY_NONE meaningfully different from grpc::InsecureChannelCredentials?

No, they are meaningfully same.

**Pull request closed without merging**